11_developer_primer/3_developing_for_dynamo/advanced-dynamo-node-customisation.md

```diff
@@ -0,0 +1,186 @@
+## Generating Custom Warning Messages Using OnLogWarningMessage <a href="#generating-custom-warning-messages-using-onlogwarningmessage" id="generating-custom-warning-messages-using-onlogwarningmessage"></a>
+
+In Dynamo, the `OnLogWarningMessage` method provides a way to log warning messages directly to Dynamo's console. This is a powerful feature, especially for Zero Touch nodes, as it allows developers to alert users when there are issues with inputs or parameters that could lead to unexpected behavior. This guide will teach you how to implement `OnLogWarningMessage` in any Zero Touch node.
+
+### Implementation Steps for `OnLogWarningMessage` <a href="#implementation-step-for-onlogwarningmessage" id="implementation-step-for-onlogwarningmessage"></a>
+
+#### Step 1: Import the Required Namespace <a href="#import-the-required-namespace" id="import-the-required-namespace"></a>
+
+`OnLogWarningMessage` is part of the `DynamoServices` namespace, so begin by adding this to your project file.
+
+```
+using DynamoServices;
+```
+
+#### Step 2: Identify When to Log Warnings <a href="#identify-when-to-log-warnings" id="identify-when-to-log-warnings"></a>
+
+Before adding a warning message, consider the logic in your method:
+
+* What conditions might cause incorrect or unexpected results?
+* Are there specific input values or parameters that the method requires to function correctly?
+
+Examples of conditions to check:
+
+* **Out-of-range values** (e.g., `if (inputValue < 0)`).
+* **Null or empty collections** (e.g., `if (list == null || list.Count == 0)`).
+* **Data type mismatches** (e.g., if a file type is unsupported).
+
+#### Step 3: Use `OnLogWarningMessage` to Log the Warning <a href="#use-onlogwarningmessage-to-log-the-warning" id="use-onlogwarningmessage-to-log-the-warning"></a>
+
+Place `OnLogWarningMessage` calls where you detect conditions that could cause issues. When the condition is met, log a warning message that provides clear guidance to the user.
+
+### Syntax for `OnLogWarningMessage` <a href="#syntax-for-onlogwarningmessage" id="syntax-for-onlogwarningmessage"></a>
+
+```
+LogWarningMessageEvents.OnLogWarningMessage("Your warning message here.");
+```
+
+### Example Implementations of `OnLogWarningMessage` <a href="#example-implementations-of-onlogwarningmessage" id="example-implementations-of-onlogwarningmessage"></a>
+
+To demonstrate `OnLogWarningMessage` in action, here are different scenarios you might encounter when building a Zero Touch node.
+
+#### Example 1: Validating Numeric Inputs <a href="#example-1-validating-numeric-inputs" id="example-1-validating-numeric-inputs"></a>
+
+In this example, we�ll build upon the custom node created in the previous �**Zero-Touch Case Study - Grid Node�;** A Method named `RectangularGrid` that generates a grid of rectangles based on `xCount` and `yCount` inputs. We will walk through testing If an input is invalid, and then using `OnLogWarningMessage` to log a warning and stop processing.
```

Comment:
A couple ? character from this sentence here to address

Reply:
Nice spot, I have updated these characters to remove the ?